src/Workspaces/Core/Portable/Workspace/Solution/DocumentState_IncrementalTreeSource.cs

```diff
@@ -0,0 +1,63 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal partial class DocumentState
+    {
+        internal class IncrementalTreeSource : ValueSource<TreeAndVersion>
+        {
+            private readonly ValueSource<TreeAndVersion> _current;
+            private readonly ValueSource<TreeAndVersion> _previous;
+ 
+            public IncrementalTreeSource(ValueSource<TreeAndVersion> current, ValueSource<TreeAndVersion> previous)
+            {
+                _current = current;
+                _previous = previous;
+            }
+
+            /// <summary>
+            /// Computes the number of incremental parses that will occur in order to realize the current source's tree.
+            /// </summary>
+            public int GetIncrementalParseDepth()
+            {
+                int depth = 0;
```

Comment:
Good point.  Please document as such.  That's a subtlety that could easily be lost.